library/core/src/num/dec2flt/parse.rs

```diff
@@ -1,121 +1,229 @@
+        negative = c == b'-';
+        if c == b'-' || c == b'+' {
+            // SAFETY: s cannot be empty
+            unsafe {
+                s.step();
+            }
+        }
     }
+    if s.first_isdigit() {
+        s.parse_digits(|digit| {
+            // no overflows here, saturate well before overflow
+            if exponent < 0x10000 {
+                exponent = 10 * exponent + digit as i64;
+            }
+        });
+        if negative { Some(-exponent) } else { Some(exponent) }
+    } else {
+        None
+    }
+}
 
-    let s = s.as_bytes();
-    let (integral, s) = eat_digits(s);
+/// Parse a partial, non-special floating point number.
+///
+/// This creates a representation of the float as the
+/// significant digits and the decimal exponent.
+fn parse_partial_number(s: &[u8], negative: bool) -> Option<(Number, usize)> {
+    let mut s = AsciiStr::new(s);
+    let start = s;
+    debug_assert!(!s.is_empty());
 
-    match s.first() {
-        None => Valid(Decimal::new(integral, b"", 0)),
-        Some(&b'e' | &b'E') => {
-            if integral.is_empty() {
-                return Invalid; // No digits before 'e'
-            }
+    // parse initial digits before dot
+    let mut mantissa = 0_u64;
+    let digits_start = s;
+    try_parse_digits(&mut s, &mut mantissa);
+    let mut n_digits = s.offset_from(&digits_start);
+
+    // handle dot with the following digits
+    let mut n_after_dot = 0;
+    let mut exponent = 0_i64;
+    let int_end = s;
+    if s.first_is(b'.') {
+        // SAFETY: s cannot be empty due to first_is
+        unsafe { s.step() };
+        let before = s;
+        try_parse_8digits(&mut s, &mut mantissa);
+        try_parse_digits(&mut s, &mut mantissa);
+        n_after_dot = s.offset_from(&before);
+        exponent = -n_after_dot as i64;
+    }
 
-            parse_exp(integral, b"", &s[1..])
+    n_digits += n_after_dot;
+    if n_digits == 0 {
+        return None;
+    }
+
+    // handle scientific format
+    let mut exp_number = 0_i64;
+    if s.first_is2(b'e', b'E') {
+        // SAFETY: s cannot be empty
+        unsafe {
+            s.step();
         }
-        Some(&b'.') => {
-            let (fractional, s) = eat_digits(&s[1..]);
-            if integral.is_empty() && fractional.is_empty() {
-                // We require at least a single digit before or after the point.
-                return Invalid;
-            }
+        // If None, we have no trailing digits after exponent, or an invalid float.
+        exp_number = parse_scientific(&mut s)?;
+        exponent += exp_number;
+    }
 
-            match s.first() {
-                None => Valid(Decimal::new(integral, fractional, 0)),
-                Some(&b'e' | &b'E') => parse_exp(integral, fractional, &s[1..]),
-                _ => Invalid, // Trailing junk after fractional part
-            }
+    // SAFETY: s must be larger than start and come from the same array
+    let len = s.offset_from(&start) as _;
```

Comment:
offset_from is safe now :)
```suggestion
    let len = s.offset_from(&start) as _;
```

Reply:
All these safety checks have been removed.